File: Data Structure/Trie/LET_no1032_Stream of Characters.cpp
class Trie {
public:
    unordered_map<char, Trie*> children;
    bool isEnd;
    
    Trie() { isEnd = false; }
    
    void insert(string& word) {
        Trie* cur = this;
        for (int i = word.size() - 1; i >= 0; i--) {
            char c = word[i];
            if (!cur->children.count(c))
                cur->children[c] = new Trie();
            cur = cur->children[c];
        }
        cur->isEnd = true;
    }
};

class StreamChecker {
private:
    vector<char> letters;
    Trie* trie;
public:
    StreamChecker(vector<string>& words) {
        trie = new Trie();
        for (string& w : words)
            trie->insert(w);
    }
    
    bool query(char letter) {
        letters.push_back(letter);
        Trie* cur = trie;
        for (int i = letters.size() - 1; i >= 0; i--) {
            char c = letters[i];
            if (cur->isEnd)
                return true;
            if (!cur->children.count(c))
                return false;
            cur = cur->children[c];
        }
        return cur->isEnd;
    }
};

/**
 * Your StreamChecker object will be instantiated and called as such:
 * StreamChecker* obj = new StreamChecker(words);
 * bool param_1 = obj->query(letter);
 */

File: Data Structure/Trie/LET_no1032_Stream of Characters.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word):
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.isEnd = True

class StreamChecker:
    def __init__(self, words: List[str]):
        self.letters = []
        self.trie = Trie()
        for w in words:
            self.trie.insert(w[::-1])

    def query(self, letter: str) -> bool:
        self.letters.append(letter)
        i = len(self.letters) - 1
        node = self.trie.root
        
        while i >= 0:
            if node.isEnd:
                return True
            if self.letters[i] not in node.children:
                return False
            node = node.children[self.letters[i]]
            i -= 1
        
        return node.isEnd


# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)